### PR TITLE
Name validation - DocDB DB+Collections, Mongo collections

### LINKED ISSUE #139: "Create New Collection" doesn't validate input?
1. Right click on a (Mongo) database and then select "Create Collection".
2. In the InputBox type in an invalid name, like "!@)*#&!@#", then hit enter.

Expected:
InputBox validation will catch invalid names.

Actual:
Nothing happens (no error message, no new collection)

### PR DESCRIPTION
**Pull request by PrashanthCorp**
Fixes #139

### CODE REVIEW

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 1]
```diff
@@ -146,6 +147,21 @@ export class MongoDatabaseTreeItem implements IAzureParentTreeItem {
 	}
 }
 
+export function validateCollectionName(collectionName: string): string | undefined | null {
+	// https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names
+	if (collectionName.length === 0) {
```
Comment (StephenWeatherford) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: if (!collectionName) ...

I've seen cases where we can get called with an undefined value and that would throw.

File: src/docdb/tree/DocDBDatabaseTreeItemBase.ts [hunk 2]
```diff
@@ -134,6 +134,16 @@ export abstract class DocDBDatabaseTreeItemBase extends DocDBTreeItemBase<Collec
         } catch (err) {
             return "Input must be a number"
         }
-        return null;
+        return undefined;
+    }
+
+    private static validateCollectionName(name: string): string | undefined | null {
+        if (name.endsWith(" ")) {
```
Comment (StephenWeatherford) [inline src/docdb/tree/DocDBDatabaseTreeItemBase.ts]: if (!name) ...

I've seen cases where we can get called with an undefined value and that would throw.

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 3]
```diff
@@ -146,6 +147,21 @@ export class MongoDatabaseTreeItem implements IAzureParentTreeItem {
...
+		return "Collection name cannot be empty";
```
Comment (StephenWeatherford) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: Does this need to be case insensitive?

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 4]
```diff
@@ -146,6 +147,21 @@ export class MongoDatabaseTreeItem implements IAzureParentTreeItem {
...
+	}
+	const systemPrefix = "system."
+	if (collectionName.startsWith(systemPrefix)) {
+		return `The "${systemPrefix}" prefix is reserved for internal use. Please refrain from using it`;
```
Comment (StephenWeatherford) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: Does this need to be case insensitive?

File: src/mongo/tree/MongoDatabaseTreeItem.ts (hunk 3 above)
Comment (PrashanthCorp) [reply src/mongo/tree/MongoDatabaseTreeItem.ts]: I don't quite get you here.

Comment (PrashanthCorp) [reply src/mongo/tree/MongoDatabaseTreeItem.ts]: If you mean the system. prefix, the link mentioned only talks about the lower case s.

File: src/docdb/tree/DocDBDatabaseTreeItemBase.ts [hunk 5]
```diff
@@ -134,6 +134,19 @@ export abstract class DocDBDatabaseTreeItemBase extends DocDBTreeItemBase<Collec
         } catch (err) {
             return "Input must be a number"
         }
-        return null;
+        return undefined;
+    }
+
+    private static validateCollectionName(name: string): string | undefined | null {
```
Comment (EricJizbaMSFT) [inline src/docdb/tree/DocDBDatabaseTreeItemBase.ts]: I don't see this being used

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 6]
```diff
@@ -146,6 +147,21 @@ export class MongoDatabaseTreeItem implements IAzureParentTreeItem {
 	}
 }
 
+export function validateCollectionName(collectionName: string): string | undefined | null {
```
Comment (EricJizbaMSFT) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: Rename this to `validateMongoCollectionName` since it's exported. We don't want to accidently use this for DocDb or something weird like that

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 7]
```diff
@@ -146,6 +147,21 @@ export class MongoDatabaseTreeItem implements IAzureParentTreeItem {
...
+	// https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names
+	if (!collectionName) {
+		return "Collection name cannot be empty";
+	}
+	const systemPrefix = "system."
+	if (collectionName.startsWith(systemPrefix)) {
+		return `The "${systemPrefix}" prefix is reserved for internal use. Please refrain from using it`;
```
Comment (EricJizbaMSFT) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: nit: I would shorten this to `"${systemPrefix}" is reserved for internal use.`

File: src/docdb/tree/DocDBDatabaseTreeItemBase.ts [hunk 8]
```diff
@@ -134,6 +134,19 @@ export abstract class DocDBDatabaseTreeItemBase extends DocDBTreeItemBase<Collec
...
+        if (!name) {
+            return "Name may not be empty";
```
Comment (EricJizbaMSFT) [inline src/docdb/tree/DocDBDatabaseTreeItemBase.ts]: nit: You're not consistent with your error messages accross docdb and mongo. I would prefer "Collection name cannot" over "Name may not" or "Name cannot"

File: src/mongo/tree/MongoDatabaseTreeItem.ts (hunk 3 above)
Comment (StephenWeatherford) [reply src/mongo/tree/MongoDatabaseTreeItem.ts]: Yes that's what I meant
